File: backend/api/models/analyze.py
import json
from typing import Optional

from pydantic import BaseModel, Field, model_validator

class AnalyzeRequest(BaseModel):
    ticket_json: dict
    limit: int = Field(15, ge=1, le=50)
    threshold: float = Field(0.40, ge=0.0, le=1.0)

    @model_validator(mode="after")
    def _check_ticket_json_size(self):
        """SEC-009: Limit ticket_json size to prevent DoS via oversized payloads."""
        serialized = json.dumps(self.ticket_json, default=str)
        if len(serialized) > 100_000:
            raise ValueError(
                f"ticket_json too large ({len(serialized)} chars, max 100,000). "
                "Reduce the payload size."
            )
        return self

class ProblemInfo(BaseModel):
    id: str
    count: int
    score: float
    is_open: bool

class FocusedPlaybookResult(BaseModel):
    playbook: str = ""
    notes: str = ""
    grounding_score: float = 0.0
    source_incident_count: int = 0
    total_similar: int = 0
    top_problem: Optional[ProblemInfo] = None
    secondary_problem: Optional[ProblemInfo] = None
    other_problems: list[str] = []
    order_ids: list[str] = []
    jira_tickets: list[str] = []

class AnalyzeResponse(BaseModel):
    """Response model for /analyze endpoint."""
    analysis_id: Optional[int] = None
    cleaned_issue: str
    confidence_score: float
    incident_exists: bool
    incident_number: Optional[str] = None
    match_count: int
    similar_incidents: list[dict]
    dominant_cluster: Optional[dict] = None
    focused_playbook: dict
    resolution_patterns: list[dict]

class AnalyzeTextRequest(BaseModel):
    text: str = Field(..., min_length=3, max_length=5000)
    limit: int = Field(15, ge=1, le=50)
    threshold: float = Field(0.40, ge=0.0, le=1.0)


class OrderAnalyzeRequest(BaseModel):
    order_number: str = Field(..., min_length=1, max_length=50)


class OrderIncidentCard(BaseModel):
    incident_number: str
    status: str = ""
    short_description: str = ""
    opened_at: Optional[str] = None
    two_line_summary: list[str] = []
    inc_tasks: list[str] = []
    alternate_orders: list[str] = []
    problem_refs: list[str] = []


class OrderAnalyzeSummary(BaseModel):
    analysis: str = ""
    accounting_actions: str = ""
    payment_activities: str = ""
    solutions: str = ""
    system_states: str = ""


class OrderAnalyzeResponse(BaseModel):
    order_number: str
    incident_count: int = 0
    message: str = ""
    incidents: list[OrderIncidentCard] = []
    summary: OrderAnalyzeSummary = Field(default_factory=OrderAnalyzeSummary)

File: frontend/src/types/analyze.ts
import type { Incident } from './incidents';

export interface KbArticle {
  sys_id: string;
  number: string;
  short_description: string;
  kb_category_display?: string;
  attached_on?: string;
  url?: string;
  has_pdf?: boolean;
  /** Set on analyze response when PDF was loaded once server-side. */
  pdf_base64?: string;
  kb_title?: string;
  source?: string;
  /** Similar-incident match % used to select this KB (0–100). */
  match_percent?: number;
  matched_via_incident?: string;
}

export interface AnalyzeResult {
  analysis_id?: number;
  cleaned_issue: string;
  confidence_score: number;
  incident_exists: boolean;
  incident_number?: string;
  match_count: number;
  similar_incidents: Incident[];
  dominant_cluster?: {
    id: number;
    cluster_name: string;
    cluster_description?: string;
    incident_count: number;
    dominant_category?: string;
    avg_resolution_hours?: number;
    avg_internal_similarity?: number;
  };
  focused_playbook: {
    playbook: string;
    notes: string;
    grounding_score: number;
    source_incident_count: number;
    total_similar: number;
    top_problem?: { id: string; count: number };
    secondary_problem?: { id: string; count: number };
    other_problems: string[];
    order_ids: string[];
    jira_tickets: string[];
    kb_articles?: KbArticle[];
    kb_source?: 'incident' | 'similar';
    kb_source_incident?: string;
    kb_match_percent?: number;
    /** Set when playbook text is summarized from a linked knowledge article. */
    playbook_source?: 'knowledge_article' | 'similar_incidents';
  };
  resolution_patterns: {
    incident_number: string;
    close_notes: string;
    similarity: number;
  }[];
}

export interface OrderIncidentCard {
  incident_number: string;
  status: string;
  short_description: string;
  opened_at?: string | null;
  two_line_summary: string[];
  inc_tasks: string[];
  alternate_orders: string[];
  problem_refs: string[];
}

export interface OrderAnalyzeSummary {
  analysis: string;
  accounting_actions: string;
  payment_activities: string;
  solutions: string;
  system_states: string;
}

export interface OrderAnalyzeResult {
  order_number: string;
  incident_count: number;
  message: string;
  incidents: OrderIncidentCard[];
  summary: OrderAnalyzeSummary;
}
